pkg/reconciler/pipeline/validate.go

```diff
@@ -58,6 +58,9 @@ func ValidatePipeline(pl *dfv1.Pipeline) error {
 			if v.Sink != nil || v.UDF != nil {
 				return fmt.Errorf("invalid vertex %q, only one of 'source', 'sink' and 'udf' can be specified", v.Name)
 			}
+			if len(pl.GetToEdges(v.Name)) == 0 || len(pl.GetFromEdges(v.Name)) > 0 {
+				return fmt.Errorf("invalid vertex %q, source must have at least 1 to edge and 0 from edges", v.Name)
```

Comment:
This error msg seems ambiguous. One can interpret it as source must have at least 1 to edge and any number of from edges. Consider changing to something like `must have 0 from edges and at least 1 to edge`